File: src/Layer2.cpp
/*******************************************************************************
 * This file is part of openWNS (open Wireless Network Simulator)
 * _____________________________________________________________________________
 *
 * Copyright (C) 2004-2007
 * Chair of Communication Networks (ComNets)
 * Kopernikusstr. 5, D-52074 Aachen, Germany
 * phone: ++49-241-80-27910,
 * fax: ++49-241-80-22242
 * email: info@openwns.org
 * www: http://www.openwns.org
 * _____________________________________________________________________________
 *
 * openWNS is free software; you can redistribute it and/or modify it under the
 * terms of the GNU Lesser General Public License version 2 as published by the
 * Free Software Foundation;
 *
 * openWNS is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 ******************************************************************************/

#include <DLL/Layer2.hpp>
#include <DLL/StationManager.hpp>
#include <DLL/UpperConvergence.hpp>

#include <WNS/ldk/ControlServiceInterface.hpp>
#include <WNS/ldk/ManagementServiceInterface.hpp>
#include <WNS/service/phy/ofdma/Handler.hpp>
#include <WNS/service/phy/ofdma/DataTransmission.hpp>
#include <WNS/ldk/utils.hpp>
#include <WNS/module/Base.hpp>
#include <WNS/pyconfig/Parser.hpp>

#include <sstream>

using namespace dll;

Layer2::StationIDType Layer2::invalidStationID = -1;
Layer2::StationIDType Layer2::broadcastStationID = -2;

Layer2::Layer2(wns::node::Interface* _node, const wns::pyconfig::View& _config, StationManager* _sm) :
	wns::node::component::Component(_node, _config),

	config(_config),
	logger(_config.get("logger")),
	stationManager( (_sm == NULL) ? TheStationManager::getInstance() : _sm),
	stationID(config.get<StationIDType>("stationID")),
	type(StationTypes::fromString(config.get<std::string>("stationType"))),
	address(config.get<wns::service::dll::UnicastAddress>("address")),
	ring(config.get<uint32_t>("ring")),
	air(),
	agent()
{
}


void
Layer2::doStartup()
{
	{
		if (!agent.knows("MAC.Id"))
		{ // new method for ProbeBus
			agent.setContext("MAC.Id", stationID);
			agent.setContext("MAC.Ring", ring);
			agent.setContext("MAC.StationType", StationTypes::fromString(config.get<std::string>("stationType")));

			wns::probe::bus::ContextProviderCollection& collection =
				getNode()->getContextProviderCollection();

			collection.addProvider(wns::probe::bus::contextprovider::Container(&this->agent));
		}
	}

	//create the DLL-FUN
	wns::pyconfig::View funConfig(config, "fun");
	fun = new wns::ldk::fun::Main(this, config);
	wns::ldk::configureFUN(fun, funConfig);

	// fire up control and management Services
	{ // do control services
		for (int ii = 0; ii<config.len("controlServices"); ++ii){
			wns::pyconfig::View controlServiceView = config.get("controlServices",ii);
			std::string serviceName = controlServiceView.get<std::string>("serviceName");
			std::string creatorName = controlServiceView.get<std::string>("nameInServiceFactory");
			wns::ldk::ControlServiceCreator* serviceCreator = wns::ldk::ControlServiceFactory::creator(creatorName);
			wns::ldk::ControlServiceInterface* service = serviceCreator->create(getCSR(), controlServiceView);
			addControlService(serviceName, service);
			MESSAGE_BEGIN(NORMAL, logger, m, getName());
			m << " Registered Control Service: " << serviceName;
			MESSAGE_END();
		}
		// do management services @todo pab: get rid of copy and paste patterns
		for (int ii = 0; ii<config.len("managementServices"); ++ii){
			wns::pyconfig::View managementServiceView = config.get("managementServices",ii);
			std::string serviceName = managementServiceView.get<std::string>("serviceName");
			std::string creatorName = managementServiceView.get<std::string>("nameInServiceFactory");
			wns::ldk::ManagementServiceCreator* serviceCreator = wns::ldk::ManagementServiceFactory::creator(creatorName);
			wns::ldk::ManagementServiceInterface* service = serviceCreator->create(getMSR(), managementServiceView);
			addManagementService(serviceName, service);
			MESSAGE_BEGIN(NORMAL, logger, m, getName());
			m << " Registered Management Service: " << serviceName;
			MESSAGE_END();
		}
	}

	/**
	 * @todo pab: when WiMAC conforms to control/management service concept
	 * enable to call fun->onFUNCreated(); */
	// resolve intra-FUN dependencies
	// fun->onFUNCreated();

	// global station registry
	stationManager->registerStation(stationID, address, this);

	if(!config.isNone("upperConvergenceName"))
	{
		std::string upperConvergenceName = config.get<std::string>("upperConvergenceName");
		UpperConvergence* upperConvergence =
			getFUN()->findFriend<UpperConvergence*>(upperConvergenceName);

		// register UpperConvergence as the DLL DataTransmissionService
		addService(config.get<std::string>("dataTransmission"), upperConvergence);
		addService(config.get<std::string>("notification"), upperConvergence);
		addService(config.get<std::string>("flowEstablishmentAndRelease"), upperConvergence);
		upperConvergence->setMACAddress(address);
	}

} // Layer2

Layer2::~Layer2()
{
    if (fun != NULL) {
	delete fun;
	fun = NULL;
    }
} // ~Layer2

Layer2::StationIDType
Layer2::getID() const
{
	return stationID;
} // getID

wns::service::dll::UnicastAddress
Layer2::getDLLAddress() const
{
	return address;
}

// pab: 2006-07-21 This function is to vanish soon.
std::string
Layer2::getName() const
{
	return getNode()->getName();
} // getName

Layer2::StationType
Layer2::getStationType() const
{
	return type;
} // getType

StationManager*
Layer2::getStationManager()
{
	return stationManager;
} // getStationManager

void
Layer2::addAssociationInfoProvider(const std::string& id, dll::services::control::AssociationInfo* provider)
{
	if (! air.knows(id) )
	{
		AssociationInfoContainer tmp;
		air.insert(id, tmp);
	}

	AssociationInfoContainer tmp2(air.find(id));
	tmp2.push_back(provider);
	air.update(id, tmp2);
}

const Layer2::AssociationInfoContainer&
Layer2::getAssociationInfoProvider(const std::string& id) const
{
	return air.find(id);
}

void
Layer2::ContextAgent::doVisit(wns::probe::bus::IContext& context) const
{
	ContextIterator iter = myContextInfo.begin();
	ContextIterator end  = myContextInfo.end();

	for( ; iter != end; ++iter)
	{
		context.insertInt( iter->first, iter->second );
	}
}

File: PyConfig/dll/Layer2.py
###############################################################################
# This file is part of openWNS (open Wireless Network Simulator)
# _____________________________________________________________________________
#
# Copyright (C) 2004-2007
# Chair of Communication Networks (ComNets)
# Kopernikusstr. 5, D-52074 Aachen, Germany
# phone: ++49-241-80-27910,
# fax: ++49-241-80-22242
# email: info@openwns.org
# www: http://www.openwns.org
# _____________________________________________________________________________
#
# openWNS is free software; you can redistribute it and/or modify it under the
# terms of the GNU Lesser General Public License version 2 as published by the
# Free Software Foundation;
#
# openWNS is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
# A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
# details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
###############################################################################

from openwns.pyconfig import attrsetter

import openwns.node
import openwns.logger
class LinkHandler:

    type = "wns.ldk.SimpleLinkHandler"

    isAcceptingLogger = None
    sendDataLogger = None
    wakeupLogger = None
    onDataLogger = None

    traceCompoundJourney = True

    def __init__(self, parentLogger = None):
        self.isAcceptingLogger = openwns.logger.Logger("WNS", "LinkHandler", True, parentLogger)
        self.sendDataLogger = openwns.logger.Logger("WNS", "LinkHandler", True, parentLogger)
        self.wakeupLogger = openwns.logger.Logger("WNS", "LinkHandler", True, parentLogger)
        self.onDataLogger = openwns.logger.Logger("WNS", "LinkHandler", True, parentLogger)

class Layer2( openwns.node.Component ):
    __associationID = -1

    # station ID, user must make sure this is set uniquely. Used for Probes Access Control
    stationID = None

    # a list of associations
    associations = None

    # String, can be "BS", "FRS" or "UT"
    stationType = None

    # an integer to denote how far away from the BS we are
    ring = None

    # Configuration of the FUN residing in the Layer2
    fun = None

    # Services to communicate with the PHY
    phyDataTransmission = None
    phyNotification = None
    phyMeasurements = None

    # Services offered to the NL
    dataTransmission = None
    notification = None

    # Services offered to the TL
    flowHandler = None
    flowEstablishmentAndRelease = None

    # MAC Adress
    address = None

    # Control Services
    controlServices = None

    # Management Services
    managementServices = None

    # Logger
    logger = None

    linkHandler = None

    # Name of the UpperConvergence FU
    upperConvergenceName = None

    #
    # layer2 Constructor
    #
    def __init__(self, node, stationName, parentLogger = None, **kw):
        self.nameInComponentFactory = None
        super(Layer2,self).__init__(node, stationName)
        self.dataTransmission = stationName + ".dllDataTransmission"
        self.notification = stationName + ".dllNotification"

        self.flowHandler = stationName + ".dllFlowHandler"
        self.flowEstablishmentAndRelease = stationName + ".dllFlowEstablishmentAndRelease"

        self.controlServices = []
        self.managementServices = []
        self.logger = openwns.logger.Logger("DLL", "L2", True, parentLogger)
        self.linkHandler = LinkHandler(self.logger)
        self.upperConvergenceName = 'upperConvergence'
        attrsetter(self, kw)

    def setPhyDataTransmission(self, serviceName):
        self.phyDataTransmission = serviceName

    def setPhyNotification(self, serviceName):
        self.phyNotification = serviceName

    def getAssociationID(self):
        self.__associationID += 1
        return self.__associationID

    def setStationID(self, number):
        if self.stationID is not None: raise AssertionError, "Do you really want to re-set the stationID?"
        self.stationID = number
        self.address = self.stationID

    def setRing(self, _ring):
        if self.ring is not None: raise AssertionError, "Do you really want to re-set the ring?"
        self.ring = _ring

    def setStationType(self, _stationType):
        if self.stationType is not None: raise AssertionError, "Do you really want to re-set the stationType?"
        self.stationType = _stationType
